### PR TITLE
Network command module for Brocade IronWare routers

### PR DESCRIPTION
**Pull request by paulquack**
##### SUMMARY

Network command module for Brocade IronWare routers
(Only one module per PR, config and facts modules to follow)

##### ISSUE TYPE

 - New Module Pull Request

##### COMPONENT NAME

- ironware_command

##### ANSIBLE VERSION

```
ansible 2.4.1.0 (ironware-module fc9bda3574) last updated 2017/10/06 12:53:41 (GMT +1100)
  config file = /Users/paul.baker/.ansible.cfg
  configured module search path = [u'/Users/paul.baker/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/paul.baker/ansible/lib/ansible
  executable location = ./ansible/bin/ansible
  python version = 2.7.13 (default, Dec 18 2016, 07:03:39) [GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.42.1)]
```

##### ADDITIONAL INFORMATION

added ironware to NETWORK_GROUP_MODULES in config/base.yml

### CODE REVIEW

File: lib/ansible/module_utils/ironware.py [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+
+from ansible.module_utils._text import to_text
+from ansible.module_utils.basic import env_fallback, return_values
+from ansible.module_utils.network_common import to_list, EntityCollection
+from ansible.module_utils.connection import Connection, exec_command
+
+_DEVICE_CONFIGS = {}
+_CONNECTION = None
+
+ironware_argument_spec = {
```
Comment (ganeshrn) [inline lib/ansible/module_utils/ironware.py:39]: In 2.5 all the top level connection arguments are deprecated and are added as sub-options of `provider` option. You can refer [here](https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/ios.py#L45) for more details.

Similar to below spec
```
ironware_provider_spec = {
    'host': dict(),
    'port': dict(type='int'),
    'username': dict(fallback=(env_fallback, ['ANSIBLE_NET_USERNAME'])),
    'password': dict(fallback=(env_fallback, ['ANSIBLE_NET_PASSWORD']), no_log=True),
    'ssh_keyfile': dict(fallback=(env_fallback, ['ANSIBLE_NET_SSH_KEYFILE']), type='path'),
    'authorize': dict(fallback=(env_fallback, ['ANSIBLE_NET_AUTHORIZE']), type='bool'),
    'auth_pass': dict(fallback=(env_fallback, ['ANSIBLE_NET_AUTH_PASS']), no_log=True),
    'timeout': dict(type='int')
}
ironware_argument_spec = {
    'provider': dict(type='dict', options=ios_provider_spec),
}
```

File: lib/ansible/module_utils/ironware.py [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+}
+
+command_spec = {
+    'command': dict(key=True),
+    'prompt': dict(),
+    'answer': dict()
+}
+
+
+def get_argspec():
```
Comment (ganeshrn) [inline lib/ansible/module_utils/ironware.py:59]: Please rename it as `get_provider_argspec()` to be in sync with other platforms

File: lib/ansible/module_utils/ironware.py [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+    return ironware_argument_spec
+
+
+def check_args(module):
+    provider = module.params['provider'] or {}
+
+    for key in ironware_argument_spec:
```
Comment (ganeshrn) [inline lib/ansible/module_utils/ironware.py:66]: This extra check is not required after above provider spec changes

File: lib/ansible/module_utils/ironware.py [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+        if key not in ['passwords', 'provider', 'authorize'] and module.params[key]:
+            module.warn('argument %s has been deprecated and will be removed in a future version' % key)
+
+    if provider:
```
Comment (ganeshrn) [inline lib/ansible/module_utils/ironware.py:70]: This check is not required after above `provider` spec change as `no_log=True` in auth_pass and passwords sub-options will take care to mask password.

File: lib/ansible/plugins/action/ironware.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+    def run(self, tmp=None, task_vars=None):
+
+        if self._play_context.connection != 'local':
+            return dict(
+                failed=True,
+                msg='invalid connection specified, expected connection=local, '
+                    'got %s' % self._play_context.connection
+            )
+
+        provider = self.load_provider()
```
Comment (ganeshrn) [inline lib/ansible/plugins/action/ironware.py:51]: Please use `load_provider()` defined [here](https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/network_common.py#L340). You can refer [this](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/ios.py#L48) action plugin.

File: lib/ansible/plugins/action/ironware.py [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+                    'msg': 'unable to open shell. Please see: ' +
+                           'https://docs.ansible.com/ansible/network_debug_troubleshooting.html#unable-to-open-shell'}
+
+        task_vars['ansible_socket'] = socket_path
+
+        result = super(ActionModule, self).run(tmp, task_vars)
+
+        return result
+
+    def load_provider(self):
```
Comment (ganeshrn) [inline lib/ansible/plugins/action/ironware.py:82]: Not required after above `load_provider()` change

File: lib/ansible/plugins/action/ironware.py [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+            if key != 'provider' and key not in provider:
+                if key in self._task.args:
+                    provider[key] = self._task.args[key]
+                elif 'fallback' in value:
+                    provider[key] = self._fallback(value['fallback'])
+                elif key not in provider:
+                    provider[key] = None
+        return provider
+
+    def _fallback(self, fallback):
```
Comment (ganeshrn) [inline lib/ansible/plugins/action/ironware.py:94]: Same as above

**Review by ganeshrn — CHANGES_REQUESTED** [review]
Please add unit test cases for this module [here](https://github.com/ansible/ansible/tree/devel/test/units/modules/network)

**Comment by paulquack** [comment]
Thanks @ganeshrn 

In addition to your suggestions, I think I probably should be submitting this module PR against devel branch rather than stable-2.4 ?

**Comment by ganeshrn** [comment]
@paulquack Since this is a new module, please raise a PR against devel branch. This will be available in 2.5 release and not required to backport to stable-2.4

**Comment by paulquack** [comment]
Ouch - sorry everyone!

**Pull request closed without merging**